fms/models/gpt_bigcode.py

```diff
@@ -0,0 +1,324 @@
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value_states: Optional[
+            Tuple[
+                torch.FloatTensor,
+            ]
+        ] = None,
+        use_cache: bool = False,
+        attn_algorithm: Optional[str] = None,
+    ):
+        # Embed the given vocabulary indices using the given attention mask, with pre-/post-norm and dropout as specified
+        # x_in: batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+
+        qlen = x.size(1)
+        klen = x.size(1)
+
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [None for _ in range(len(self.layers))]
+
+        # if we are using the cache, the key length needs to be extended with the past keys length
+        if use_cache and past_key_value_states[0] is not None:
+            klen += past_key_value_states[0][0].size(-2)
+
+        # if mask is none, we need to compute mask
+        is_causal_mask = False
+        if mask is None:
+            if x is None:
+                raise ValueError("cannot create a mask when x is None")
+            # we are caching and can assume all 1s in the mask
+            if use_cache and klen != 1 and qlen == 1:
+                # b x h x qlen x kvlen
+                mask = torch.ones(qlen, klen, device=x.device)
+            else:
+                pad_id: int = self.config.pad_id
+                is_pad: torch.BoolTensor = x == pad_id
+                mask: torch.BoolTensor = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+                mask = mask.tril(diagonal=0)
+
+        x_emb = self.embedding(x)
+
+        if position_ids is None:
+            _position_ids = torch.arange(
+                0, qlen, dtype=torch.long, device=x.device
+            ).repeat(x.size(0), 1)
+            # Compute position_ids based on cache config
+            if use_cache and past_key_value_states[0] is not None:
+                _position_ids += past_key_value_states[0][0].shape[2]
+        else:
+            # use the position ids provided by the user directly
+            _position_ids = position_ids
+
+        if self.config.pad_id is not None:
+            is_pad = x == self.config.pad_id
+
+            # if position_ids were not given by the user, correct for pads
+            # assume position_ids is already corrected for pads if given by the user
+            if position_ids is None:
+                _position_ids = _position_ids.sub(is_pad.cumsum(1))
+                # In case of left-padding, prevent negative indices (get zeroed anyway)
+                _position_ids = _position_ids.clamp(min=0)
+
+            # zero out the associated position embeddings
+            position_out = self.position_embedding(_position_ids).mul(
+                ~is_pad.unsqueeze(-1)
```

Comment:
wouldn't setting `padding_idx` on position_embedding accomplish this more easiliy?

Reply:
Looking into this, I don’t think `padding_idx` here helps, might not need it on the embedding either. What we are doing here is if the input includes pads, we zero out those positions. If we wanted to use padding_idx properly (which we could, but it would not be using pad_id directly), we would need to set padding_idx to some other value (when pad_id exists), then make sure the position_ids are set to that padding_idx value when a pad exists at that position in the input. I don’t think its worth doing that here as we still would need to check for is_pad, then modify position_ids.

Reply:
Agreeing with @JRosenkranz - some HF models actually do use `padding_idx` in their position embedding tensors, and then increment the count from there. So if `padding_idx = 4`, the position embeddings are `[null, null, null, null, pad, pos0, pos1, pos2...]`. This does skip the zeroing step, but still involves a manual check and correction for pad tokens, and also means that the position embedding tensor can have length up to `vocab_size + max_len`, depending on your `padding_idx` value. This is bad because in general, `vocab_size >> max_len`. @JRosenkranz 's hypothetical of untying the `padding_idx` values across the position and vocab embeddings runs into similar issues - it still needs manual position index correction, and the position embedding tensor is still larger than `max_len`, which is unintuitive for the user/designer.

Reply:
let me see if I understand this correctly:

for this to make sense with `nn.Embedding(... padding_idx = pad_idx)`, we'd need to set `position_ids = torch.where(is_pad, padding_idx, position_ids)` ? Otherwise when padding_idx != 0, it wouldn't properly zero out the output?

Reply:
For your first question, you're correct (that's what I meant by manual check/correction), though that's not sufficient on its own. Not sure I follow the second question. The issue is a pigeonhole problem - if there are n embeddings for n positions, then none of them are available for the pad token's zero vector. So either you dedicate a position index to the pad token, and then count "around" that value for all your actual tokens, or you do the zeroing outside the vocab tensor.